### PR TITLE
Tn share image part3

### PR DESCRIPTION
**Pull request by justinpearsonappfolio**
Hi Haomin, would you mind reviewing this one last PR for us? We refactored our JS from issue #7 Part 2 after some feedback from Ramon and DDM, and it would be interesting to hear your feedback too. Thanks very much!

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -43,6 +43,7 @@ end
 group :test do
   gem 'ae_page_objects'
   gem 'capybara'
+  gem 'pry'
```
Comment (HaominZhang) [inline Gemfile]: I think it would be better that we could put gem installation into its own commit

File: app/assets/javascripts/share_image/share_image.js [hunk 2]
```diff
@@ -3,50 +3,66 @@ class ShareModal {
...
-      this.status_code = xhr.status;
+      const message = xhr.responseJSON.message;
 
       if (xhr.status == 400) {
-        this.modal_html = xhr.responseJSON.modal_html;
-        $(this.$modal).modal('hide');
+        if ($('.text-danger').length > 0) {
+          $('.text-danger').remove();
+        }
+        $('#share_form_email_address').after(this.errorMessage(message));
```
Comment (HaominZhang) [inline app/assets/javascripts/share_image/share_image.js]: Correct me if I am wrong. I think this `id` is generated by simple form "implicitly" (we didn't define it). So if in the future anybody refactored the form, adding their own `id` or something else, then this javascript may break without any recognition. My proposal (for Ropes and also for future work in APM) is having a customized `class` on the field and we can use the class here as the css selector. By doing this, it is easy for global search and future update. Does it make sense to you?

File: app/assets/javascripts/share_image/share_image.js [hunk 3]
```diff
@@ -3,50 +3,66 @@ class ShareModal {
...
+    return '<div class="alert alert-success">' + message + '</div>';
+  }
+
+  flashWarningMessage(message) {
+    return '<div class="alert alert-warning">' + message + '</div>';
   }
 }
 
-new ShareModal('.js-share-button', '.js-share-modal', '#ShareModal');
+new ShareModal('.js-share-modal', '#ShareModal');
```
Comment (HaominZhang) [inline app/assets/javascripts/share_image/share_image.js]: Comparing to my above code review, I think we could still use `id` here because this `id` is defined by developers

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -68,16 +62,20 @@ def email_image
...
 
   def send_email
+    url = Image.find(share_params[:image_id]).url
     ImageSharerMailer.image_sharer_email(share_params[:email_address],
-                                         msg: share_params[:message], url: @im.url).deliver
+                                         msg: share_params[:message], url: url).deliver
     render json: { message: 'Image shared!' }, status: :ok
-  rescue Net::SMTPAuthenticationError
-    render json: { message: 'SendGrid login error!' }, status: :unprocessable_entity
+  rescue Net::SMTPAuthenticationError, ActiveRecord::RecordNotFound
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: Minor: Could we separate the rescue of these two different types of exceptions. I think the `RecordNotFound` exception comes from `Image.find` right? Which means, the image that the user trying to share doesn't exist, which sounds like a user-side error. However, the `SMTPAuthenticationError` has nothing to do with user's behavior. So ideally we could let user know a more specific error so they know what has gone wrong. How do you guys think?

File: app/views/images/_hidden_share_modal.html.erb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<div class="js-share-modal">
```
Comment (HaominZhang) [inline app/views/images/_hidden_share_modal.html.erb]: I don't think `hidden_share_modal` is a good name for this file. Although it is initially hidden, it will finally show up. But if we just look at the name, it may be confusing because others could take it as a forever hidden modal. Do you mind renaming this file? I think just `share_modal` is fine

File: Gemfile (hunk 1 above)
Comment (justinpearsonappfolio) [reply Gemfile]: Sounds good!

File: app/assets/javascripts/share_image/share_image.js (hunk 2 above)
Comment (justinpearsonappfolio) [reply app/assets/javascripts/share_image/share_image.js]: Thanks -- we'll use a custom class.

File: app/assets/javascripts/share_image/share_image.js (hunk 3 above)
Comment (justinpearsonappfolio) [reply app/assets/javascripts/share_image/share_image.js]: Sounds good

**Comment by justinpearsonappfolio** [comment]
Hi Haomin, we made the last couple changes you suggested -- would you mind looking it over one last time? Thanks!

File: app/assets/javascripts/share_image/share_image.js [hunk 6]
```diff
@@ -3,50 +3,73 @@ class ShareModal {
     this.$shareButton = shareButtonClass;
     this.$modalParent = modalParentClass;
     this.$modal = modalId;
-    this.status_code = -1;
+
+    console.log('Inside ShareModal constructor!');
```
Comment (HaominZhang) [inline app/assets/javascripts/share_image/share_image.js]: Did we forget to remove these two lines?

**Comment by justinpearsonappfolio** [comment]
Thanks Haomin! I think we are finally FINISHED with the Ropes Image Sharing project!
